File: 22_专题/前缀表达式(前缀和)/全部元素都不大于 k的子数组个数/795. 区间子数组个数.ts
import { betweenK } from '../api'

/**
 * @param {number[]} nums
 * @param {number} left
 * @param {number} right
 * @return {number}
 * 求连续、非空且其中最大元素满足大于等于L 小于等于R的子数组个数。
 */
const numSubarrayBoundedMax = function (nums: number[], left: number, right: number): number {
  return betweenK(left, right, nums)
}

console.log(numSubarrayBoundedMax([2, 1, 4, 3], 2, 3))

File: 16_滑动窗口/795. 区间子数组个数.py
# 795. 区间子数组个数
# https://leetcode.cn/problems/number-of-subarrays-with-bounded-maximum/description/
# 类似 https://leetcode.cn/problems/count-subarrays-with-fixed-bounds/
#
# 给你一个整数数组 nums 和两个整数：left 及 right 。
# 找出 nums 中连续、非空且其中最大元素在范围 [left, right] 内的子数组，并返回满足条件的子数组的个数。

from typing import List


class Solution:
    def numSubarrayBoundedMax(self, nums: List[int], left: int, right: int) -> int:
        res = 0
        pos1, pos2 = -1, -1
        for i, v in enumerate(nums):
            if v > right:
                pos1 = i
            if v >= left:
                pos2 = i
            res += pos2 - pos1
        return res
